From e8f1e99c96a8303421e66b0feda1651a11c8b250 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Tim=20R=C3=BChsen?= <tim.ruehsen@gmx.de>
Date: Mon, 20 May 2024 13:19:06 +0200
Subject: [PATCH] Accept --progress=dot:... for backwards compatibility

* src/options.c (parse_progress_type): Fix checking dot options.
* tests/test-wget-1.c: Add check for --progress variants.
---
 src/options.c       |  6 +++---
 tests/test-wget-1.c | 10 ++++++++++
 2 files changed, 13 insertions(+), 3 deletions(-)

diff --git a/src/options.c b/src/options.c
index 7684b795..0f7b3f35 100644
--- a/src/options.c
+++ b/src/options.c
@@ -798,13 +798,13 @@ static int WGET_GCC_PURE WGET_GCC_NONNULL((1)) parse_progress_type(option_t opt,
 
 	if (!wget_strcasecmp_ascii(val, "none"))
 		*((char *)opt->var) = PROGRESS_TYPE_NONE;
-	else if (!wget_strncasecmp_ascii(val, "bar", 3)) {
+	else if (!wget_strncasecmp_ascii(val, "bar", 3) && (val[3] == ':' || val[3] == 0)) {
 		*((char *)opt->var) = PROGRESS_TYPE_BAR;
 		// Silent Wget compatibility
-		if (!wget_strncasecmp_ascii(val+3, ":force", 6) || !wget_strncasecmp_ascii(val+3, ":noscroll:force", 15)) {
+		if (!wget_strncasecmp_ascii(val+4, "force", 5) || !wget_strncasecmp_ascii(val+4, "noscroll:force", 14)) {
 			config.force_progress = true;
 		}
-	} else if (!wget_strcasecmp_ascii(val, "dot")) {
+	} else if (!wget_strncasecmp_ascii(val, "dot", 3) && (val[3] == ':' || val[3] == 0)) {
 		// Wget compatibility, whether want to support 'dot' depends on user feedback.
 		info_printf(_("Progress type '%s' ignored. It is not implemented yet\n"), val);
 	} else {
diff --git a/tests/test-wget-1.c b/tests/test-wget-1.c
index fdd4f54e..8a08d74d 100644
--- a/tests/test-wget-1.c
+++ b/tests/test-wget-1.c
@@ -626,6 +626,16 @@ int main(void)
 			{	NULL } },
 		0);
 
+	// test different --progress options to be accepted
+	wget_test(
+		WGET_TEST_OPTIONS, "--progress=none --progress=bar --progress=bar:force --progress=bar:noscroll:force --progress=dot --progress=dot:giga",
+		WGET_TEST_REQUEST_URL, "dummy.txt",
+		WGET_TEST_EXPECTED_ERROR_CODE, 0,
+		WGET_TEST_EXPECTED_FILES, &(wget_test_file_t []) {
+			{	"dummy.txt", urls[3].body },
+			{	NULL } },
+		0);
+
 	// test--https-only
 	wget_test(
 		WGET_TEST_OPTIONS, "--https-only -r -nH",
-- 
2.43.0

